example/example/spiders/books_06.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+"""
+Scrapy spider with reusable components used for two types of pages:
+* Listings page, with the components: book list and pagination
+* Detail book page, with the components: the item itself, the breadcrumbs and
+  the recently viewed list of books
+
+All links to detail book pages are found are followed.
+All links to search result pages are followed as well.
+"""
+import scrapy
+import attr
+
+from scrapy_po import WebPage
+
+
+class ListingsPiece(WebPage):
```

Comment:
Any ideas about a better API? Inheriting "page part" component from a WebPage looks a bit strange, on a first sight

Reply:
Yes is strange. Also, all the "pieces" are incorporating the selector API without sharing it, which I don't like very much. Well, I have to think about it. This is very preliminary to me, I just noticed that I needed the recursive injection and started to experiment with it.

Reply:
> Also, all the "pieces" are incorporating the selector API without sharing it, which I don't like very much.

Sorry, could you please explain it more?

Reply:
If the goal of inheriting is to be able to say "please do dependency injection for this class", it could make sense to have a base class for this (e.g. Component, or something like this - this is becoming Java pretty fast), and have PageObject inherit from Component.

Reply:
Aha, maybe I know what you mean. In the code I was showing you on a call (which is not commited), selector-specific logic is moved to a mixin:

```py
class PageObject(abc.ABC):
    """
    All web page objects should inherit from this class.
    It is a marker for a middleware to pick anargument and populate it.

    Instead of inheriting you can also use ``PageObject.register(MyWebPage)``.
    """
    pass


class ItemPage(PageObject):
    """ Page Object which requires to_item method to be implemented. """
    @abc.abstractmethod
    def to_item(self):
        pass


class ResponseShortcutsMixin:
    """
    A mixin with common shortcut methods for working with self.response.
    It requires "response" attribute to be present.
    """
    _cached_selector = None

    @property
    def selector(self) -> parsel.Selector:
        from parsel.selector import Selector
        if self._cached_selector is None:
            self._cached_selector = Selector(self.html)
        return self._cached_selector

    @property
    def url(self):
        return self.response.url

    @property
    def html(self):
        return self.response.html

    def xpath(self, query, **kwargs):
        return self.selector.xpath(query, **kwargs)

    def css(self, query):
        return self.selector.css(query)


@attr.s(auto_attribs=True)
class WebPage(PageObject, ResponseShortcutsMixin):
    """
    Default WebPage class. It uses basic response data (html, url),
    and provides xpath / css shortcuts.

    This class is most useful as a base class for page objects, when
    extraction logic is similar to what's usually happening in scrapy
    spider callbacks.
    """
    response: ResponseData
```

Let me asend a PR with these changes.

Reply:
https://github.com/scrapinghub/scrapy-po/pull/3

Reply:
Thanks @kmike ! I left some comments in the PR itself that I think covers yoursquestions.